prov/util/src/util_av.c

```diff
@@ -0,0 +1,957 @@
+		    const void *dest_addr, size_t dest_addrlen,
+		    void **src_addr, size_t *src_addrlen)
+{
+	switch (addr_format) {
+	case FI_SOCKADDR:
+	case FI_SOCKADDR_IN:
+	case FI_SOCKADDR_IN6:
+		return fi_get_src_sockaddr(dest_addr, dest_addrlen,
+					   (struct sockaddr **) src_addr,
+					   src_addrlen);
+	default:
+		return -FI_ENOSYS;
+	}
+}
+
+static int fi_get_sockaddr(int sa_family, uint64_t flags,
+			   const char *node, const char *service,
+			   struct sockaddr **addr, size_t *addrlen)
+{
+	struct addrinfo hints, *ai;
+	int ret;
+
+	memset(&hints, 0, sizeof hints);
+	hints.ai_family = sa_family;
+	hints.ai_socktype = SOCK_STREAM;
+	if (flags & FI_SOURCE)
+		hints.ai_flags = AI_PASSIVE;
+
+	ret = getaddrinfo(node, service, &hints, &ai);
+	if (ret)
+		return -FI_ENODATA;
+
+	*addr = mem_dup(ai->ai_addr, ai->ai_addrlen);
+	if (!*addr) {
+		ret = -FI_ENOMEM;
+		goto out;
+	}
+
+	*addrlen = ai->ai_addrlen;
+out:
+	freeaddrinfo(ai);
+	return ret;
+}
+
+int fi_get_addr(uint32_t addr_format, uint64_t flags,
+		const char *node, const char *service,
+		void **addr, size_t *addrlen)
+{
+	switch (addr_format) {
+	case FI_SOCKADDR:
+		return fi_get_sockaddr(0, flags, node, service,
+				       (struct sockaddr **) addr, addrlen);
+	case FI_SOCKADDR_IN:
+		return fi_get_sockaddr(AF_INET, flags, node, service,
+				       (struct sockaddr **) addr, addrlen);
+	case FI_SOCKADDR_IN6:
+		return fi_get_sockaddr(AF_INET6, flags, node, service,
+				       (struct sockaddr **) addr, addrlen);
+	default:
+		return -FI_ENOSYS;
+	}
+}
+
+static void *util_av_get_data(struct util_av *av, int index)
+{
+	return (char *) av->data + (index * av->addrlen);
+}
+
+void *fi_av_get_addr(struct util_av *av, int index)
+{
+	return util_av_get_data(av, index);
+}
+
+static void util_av_set_data(struct util_av *av, int index,
+			     const void *data, size_t len)
+{
+	memcpy(util_av_get_data(av, index), data, len);
+}
+
+static int fi_verify_av_insert(struct util_av *av, uint64_t flags)
+{
+	if ((av->flags & FI_EVENT) && !av->eq) {
+		FI_WARN(av->prov, FI_LOG_AV, "no EQ bound to AV\n");
+		return -FI_ENOEQ;
+	}
+
+	if (flags & ~(FI_MORE)) {
+		FI_WARN(av->prov, FI_LOG_AV, "unsupported flags\n");
+		return -FI_ENOEQ;
+	}
+
+	return 0;
+}
+
+/*
+ * Must hold AV lock
+ */
+static int util_av_hash_insert(struct util_av_hash *hash, int slot, int index)
+{
+	int entry, i;
+
+	if (slot < 0 || slot >= hash->slots)
+		return -FI_EINVAL;
+
+	if (hash->table[slot].index == UTIL_NO_ENTRY) {
+		hash->table[slot].index = index;
+		return 0;
+	}
+
+	if (hash->free_list == UTIL_NO_ENTRY)
+		return -FI_ENOSPC;
+
+	entry = hash->free_list;
+	hash->free_list = hash->table[hash->free_list].next;
+
+	for (i = slot; hash->table[i].next != UTIL_NO_ENTRY; )
+		i = hash->table[i].next;
+
+	hash->table[i].next = entry;
+	hash->table[entry].index = index;
+	hash->table[entry].next = UTIL_NO_ENTRY;
+	return 0;
+}
+
+static int fi_av_insert_addr(struct util_av *av, const void *addr, int slot,
```

Comment:
(I'm so late to commenting on this PR because I was on vacation for a couple of weeks, sorry for the delay)

It would be nice if the `fi_` prefix were largely reserved for user-facing types and symbols.  That is, `fi_av_insert_addr` is just a `static` helper function for the `util` provider's AV implementation but it sounds extremely similar to the user-facing `fi_av_insert`.  Shouldn't this (and the many other `fi_`-prefixed symbols/types in this directory) just be `util_`-prefixed or maybe `fiu_`-prefixed instead?